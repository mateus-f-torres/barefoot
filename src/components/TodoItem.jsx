//@flow
import * as React from 'react';
import type {Todo} from '../types/props.js';

type Props = Todo & {
  toggleTodo: (SyntheticEvent<HTMLParagraphElement>) => void,
  deleteTodo: (SyntheticEvent<HTMLButtonElement>) => void,
};

const TodoItem = (props: Props) => (
  <li id={props.id}>
    <p data-test="todo-text" onClick={props.toggleTodo}
      className={props.completed ? 'done' : null}>
      {props.text}
    </p>
    <button data-test="todo-delete" onClick={props.deleteTodo}>delete</button>
  </li>
);

export default TodoItem;
