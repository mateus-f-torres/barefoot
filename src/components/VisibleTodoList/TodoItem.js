//@flow
import * as React from 'react';
import styled from 'styled-components';
import type {Todo} from 'Types/props.js';
import Button from 'Components/common/Button';
import trash from 'Assets/images/trash.png';

const MAIN_TEXT= '#1b1b1e';

const ListItem = styled.li`
  display: flex;
  background-color: darkgrey;
  color: ${MAIN_TEXT};
  justify-content: space-between;
  align-items: center;
  height: 45px;
  line-height: 45px;
  &:nth-child(2n) { 
    background-color: white;
    color: #666;
  }
`;

const ListItemText = styled.p`
  display: inline-block;
  width: 100%;
  padding-left: 10px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  text-decoration: ${(props) => props.done ? 'line-through' : 'none'};
  &:hover {
    cursor: pointer;
  }
`;

const ListItemIcon = styled.img`
  width: 30px;
  heigth: 30px;
  position: relative;
  bottom: 5px;
`;

type Props = Todo & {
  toggleTodo: (SyntheticEvent<HTMLParagraphElement>) => void,
  deleteTodo: (SyntheticEvent<HTMLButtonElement>) => void,
};

const TodoItem = (props: Props) => (
  <ListItem id={props.id}>
    <ListItemText
      done={props.completed ? true : false}
      data-test="todo-text"
      onClick={props.toggleTodo}>
      {props.text}
    </ListItemText>
    <Button
      data-test="todo-delete"
      onClick={props.deleteTodo}>
      <ListItemIcon src={trash}/>
    </Button>
  </ListItem>
);

export default TodoItem;
