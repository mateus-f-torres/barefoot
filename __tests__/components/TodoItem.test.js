import React from 'react';
import {shallow} from 'enzyme';
import TodoItem from '../../src/components/TodoItem.jsx';

describe('<TodoItem />', () => {
  const mockProps = {
    id: 5,
    text: 'Hello World',
    completed: false,
  };

  describe('Unit Test', () => {
    test('render props.text', () => {
      const wrapper = shallow(<TodoItem {...mockProps}/>);

      expect(wrapper.find('[data-test="todo-text"]')
        .text()).toMatch(/^Hello\sWorld$/);
    });

    test('li id matches props.id', () => {
      const wrapper = shallow(<TodoItem {...mockProps}/>);

      expect(wrapper.prop('id')).toBe(mockProps.id);
    });

    test('className === null if (props.completed === false)', () => {
      const wrapper = shallow(<TodoItem {...mockProps}/>);

      expect(wrapper.find('[data-test="todo-text"]')
        .prop('className')).toBeNull();
    });

    test('className === "done" if (props.completed === true)', () => {
      const mockCompleted = Object.assign({}, mockProps, {completed: true});
      const wrapper = shallow(<TodoItem {...mockCompleted}/>);

      expect(wrapper.find('[data-test="todo-text"]')
        .prop('className')).toMatch(/^done$/);
    });

    it('calls toggleTodo only when p is clicked', () => {
      const mockOnClick = jest.fn();
      const wrapper = shallow(
        <TodoItem toggleTodo={mockOnClick} {...mockProps}/>
      );

      wrapper.find('[data-test="todo-text"]').simulate('click');
      wrapper.find('[data-test="todo-delete"]').simulate('click');

      expect(mockOnClick.mock.calls.length).toEqual(1);
    });

    it('calls deleteTodo only when button is clicked', () => {
      const mockOnClick = jest.fn();
      const wrapper = shallow(
        <TodoItem deleteTodo={mockOnClick} {...mockProps}/>
      );

      wrapper.find('[data-test="todo-delete"]').simulate('click');
      wrapper.find('[data-test="todo-text"]').simulate('click');

      expect(mockOnClick.mock.calls.length).toEqual(1);
    });
  });

  describe('Snapshot Test', () => {
    it('matches current production snapshot', () => {
      const wrapper = shallow(<TodoItem {...mockProps}/>);

      expect(wrapper).toMatchSnapshot();
    });
  });
});
