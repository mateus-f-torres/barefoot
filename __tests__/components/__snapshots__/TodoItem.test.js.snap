// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoItem /> Snapshot Test matches current production snapshot 1`] = `
<TodoItem__ListItem
  id={5}
>
  <TodoItem__ListItemText
    data-test="todo-text"
    done={false}
  >
    Hello World
  </TodoItem__ListItemText>
  <Button
    data-test="todo-delete"
  >
    delete
  </Button>
</TodoItem__ListItem>
`;
